### PR TITLE
Bratseth/code map

### PR DESCRIPTION
**Pull request by bratseth**
@geirst please review

### CODE REVIEW

File: Code-map.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+Search container, layered on jDisc container:
+
+- [container-search](https://github.com/vespa-engine/vespa/tree/master/container-search) - Query-Result processing framework (Searchers) layered over the generic processing framework, query profiles, and the global query execution logic: Dispatch (scatter-gather), grouping, etc.
+
+Document operation modules:
+
+- [document](https://github.com/vespa-engine/vespa/tree/master/document) - The document model - documents, fields and document types, and operations on documents.
+- [messagebus](https://github.com/vespa-engine/vespa/tree/master/messagebus) - Generic async, multi-hop message passing implemented in both Java and C++.
+- [jdics_messagebus_service](https://github.com/vespa-engine/vespa/tree/master/jdisc_messagebus_service) - MessageBus connector for jDisc.
```
Comment (jonmv) [inline Code-map.md]: jdics

**Comment by jonmv** [comment]
Nice initiative :)

File: Code-map.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+jDisc core modules:
+
+- [jdisc_core](https://github.com/vespa-engine/vespa/tree/master/jdisc_core) - The core jDisc functionality
+- [jdisc_http_service](https://github.com/vespa-engine/vespa/tree/master/jdisc_http_service) - HTTP connector for jDisc, implemented using Jetty.
+
+jDisc container modules, layered on jDisc core:
+
+- [container-disc](https://github.com/vespa-engine/vespa/tree/master/container-disc) - Integration between the jDisc container and jDisc core layers.
+- [container-core](https://github.com/vespa-engine/vespa/tree/master/container-core) - Core jDisc container functionality: Metrics, OSGi integration for component bundles, etc.
+- [component](https://github.com/vespa-engine/vespa/tree/master/component) - the component model. Components (in Java) will implement or subclass a type for this module.
```
Comment (geirst) [inline Code-map.md]: Start with upper case 'T' in 'the component...' for consistency? Applies for the next three items as well

File: Code-map.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+- [indexinglanguage](https://github.com/vespa-engine/vespa/tree/master/indexinglanguage) - Implementation of the "indexing" language which is used to express the statements prefixed by "indexing:" in the search definition. 
+- [docprocs](https://github.com/vespa-engine/vespa/tree/master/docprocs) - Document processor components bundled with Vespa. Notably the Indexingprocessor - a document processor invoking the indexing language statements configured for the document type in question on document operations.
+- [vespaclient-container-plugin](https://github.com/vespa-engine/vespa/tree/master/vespaclient-container-plugin) - Implements the document/v1 API and internal API used by the Java HTTP client on top of the jDisc container, forwarding to the Document API.
+- [vespa-http-client](https://github.com/vespa-engine/vespa/tree/master/vespa-http-client) - client for fast writing to the internal API implemented by vespaclient-container-plugin.
```
Comment (geirst) [inline Code-map.md]: Uppercase 'C' in 'client'?

File: Code-map.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+## Content nodes
+
+Content nodes store all data in Vespa, maintains reverse indexes and performs the distributed parts of query execution - matching, ranking and grouping/aggregation.
+This is written in C++.
+
+- [searchcore](https://github.com/vespa-engine/vespa/tree/master/searchcore) - core functionality for maintaining indexes, matching, data storage, grouping, as well as the content node server itselv (called proton).
```
Comment (geirst) [inline Code-map.md]: Uppercase 'C' in 'core' to be consistent with the rest of the document? Applies for the entire section.

File: Code-map.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+- [searchlib](https://github.com/vespa-engine/vespa/tree/master/searchlib) - libraries invoked by searchcore: Ranking (feature execution framework (fef), rank feature implementations, ranking expressions), index and btree implementations etc. This also contains the Java libraries for ranking.
```
Comment (geirst) [inline Code-map.md]: Should mention attribute vector implementations as well.

File: Code-map.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+- [storage](https://github.com/vespa-engine/vespa/tree/master/storage/src/vespa/storage) - system for elastic and auto-recovering data storage over clusters of nodes.
+- [eval](https://github.com/vespa-engine/vespa/tree/master/eval) - library for efficient evaluation of ranking expressions.
```
Comment (geirst) [inline Code-map.md]: Could mention that tensor api/implementation is located here.

File: Code-map.md (hunk 1 above)
Comment (bratseth) [reply Code-map.md]: Fixed, thanks

File: Code-map.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bratseth) [reply Code-map.md]: Fixed, thanks

File: Code-map.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+This document aims to provide a map from the
+[functional elements](https://docs.vespa.ai/documentation/overview.html)
+of Vespa to the most important modules in the flat module structure in the
+[code base on GitHub](https://github.com/vespa-engine/vespa).
+
+It covers the modules you are most likely to encounter as a developer. 
+The rest are either small and needed for technical reasons or doing one thing 
+which should be self-explanatory, or implementing the cloud service run by the 
+Vespa team which we don't expect anybody else to run and therefor be interested
```
[the remark this answers is not in the document]
Comment (bratseth) [reply Code-map.md]: Fixed, thanks

File: Code-map.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+- [storageapi](https://github.com/vespa-engine/vespa/tree/master/storageapi/src/vespa/storageapi) - message bus messages and implementation for the document API.
+- [clustercontroller-core](https://github.com/vespa-engine/vespa/tree/master/clustercontroller-core) - cluster controller for storage, implemented in Java. This provides singular node-level decision making for storage, based on ZooKeeper.
+
+
+## Configuration and administration
+
+The third major subsystem in Vespa is responsible for managing configuration, clusters, application deployment and similar.
+It is implemented in Java.
+
+- [configserver](https://github.com/vespa-engine/vespa/tree/master/configserver) - the server to which applications are deployed, and nodes send requests for the configuration derived from these deployments.
```
Comment (geirst) [inline Code-map.md]: Uppercase 'T' in 'the'? Applies for entire section

File: Code-map.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+- [container-di](https://github.com/vespa-engine/vespa/tree/master/container-di) - component dependency injection framework for the container, compatible with Guice annotations but an separate implementation which handles injection of config and injection of component collections.
+- [processing](https://github.com/vespa-engine/vespa/tree/master/processing) - generic, chainable request-response processing framework (Processors).
+
+Search container, layered on jDisc container:
+
+- [container-search](https://github.com/vespa-engine/vespa/tree/master/container-search) - Query-Result processing framework (Searchers) layered over the generic processing framework, query profiles, and the global query execution logic: Dispatch (scatter-gather), grouping, etc.
+
+Document operation modules:
+
+- [document](https://github.com/vespa-engine/vespa/tree/master/document) - The document model - documents, fields and document types, and operations on documents.
```
Comment (geirst) [inline Code-map.md]: Mention that this is both Java and C++?

File: Code-map.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+- [indexinglanguage](https://github.com/vespa-engine/vespa/tree/master/indexinglanguage) - Implementation of the "indexing" language which is used to express the statements prefixed by "indexing:" in the search definition. 
+- [docprocs](https://github.com/vespa-engine/vespa/tree/master/docprocs) - Document processor components bundled with Vespa. Notably the Indexingprocessor - a document processor invoking the indexing language statements configured for the document type in question on document operations.
+- [vespaclient-container-plugin](https://github.com/vespa-engine/vespa/tree/master/vespaclient-container-plugin) - Implements the document/v1 API and internal API used by the Java HTTP client on top of the jDisc container, forwarding to the Document API.
+- [vespa-http-client](https://github.com/vespa-engine/vespa/tree/master/vespa-http-client) - client for fast writing to the internal API implemented by vespaclient-container-plugin.
+
+## Content nodes
+
+Content nodes store all data in Vespa, maintains reverse indexes and performs the distributed parts of query execution - matching, ranking and grouping/aggregation.
```
Comment (geirst) [inline Code-map.md]: Mention attribute vectors (forward indexes) as well?

**Review by geirst — COMMENTED** [review]
Overall looks very good. Some nitpicks to consider

File: Code-map.md (hunk 2 above)
Comment (bratseth) [reply Code-map.md]: Done (lowercase everywhere)

File: Code-map.md (hunk 9 above)
Comment (bratseth) [reply Code-map.md]: Done, thanks.

File: Code-map.md (hunk 10 above)
Comment (bratseth) [reply Code-map.md]: Done, thanks.

File: Code-map.md (hunk 6 above)
Comment (bratseth) [reply Code-map.md]: Done, thanks.

File: Code-map.md (hunk 5 above)
Comment (bratseth) [reply Code-map.md]: Done, thanks.

**Review by geirst — COMMENTED** [review]
👍

**Pull request merged**